internal/cloudfront/service.go

```diff
@@ -0,0 +1,340 @@
+// Copyright (c) 2021 GPBR Participacoes LTDA.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of
+// this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to
+// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package cloudfront
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/go-multierror"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Gympass/cdn-origin-controller/api/v1alpha1"
+	"github.com/Gympass/cdn-origin-controller/internal/config"
+	"github.com/Gympass/cdn-origin-controller/internal/k8s"
+	"github.com/Gympass/cdn-origin-controller/internal/route53"
+	"github.com/Gympass/cdn-origin-controller/internal/strhelper"
+)
+
+const (
+	reasonFailed  = "FailedToReconcile"
+	reasonSuccess = "SuccessfullyReconciled"
+)
+
+// Service handles operations involving CloudFront
+type Service struct {
+	client.Client
+
+	Config    config.Config
+	Recorder  record.EventRecorder
+	AliasRepo route53.AliasRepository
+	DistRepo  DistributionRepository
+	Fetcher   k8s.IngressFetcher
+}
+
+// Reconcile an Ingress resource of any version
+func (s *Service) Reconcile(ctx context.Context, reconciling k8s.CDNIngress, ing client.Object) error {
```

Comment:
This could have been `Sync` or similar to what we have on other places. I think it makes sense for our services to have an idempotent & declarative API, like Sync, Reconcile, etc. because breaking it down into specific operations (create, delete, update) is part of the business logic.

In a CRUD system the service operations match the interface operations. IMO, In controllers, the interface is declarative, so our services API should be too in order to correctly encapsulate business logic within it. Wdyt?